Full text search optimization

Ergus opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):

Optimize full-text-search filtering with direct index access. At the moment the only optimized function is `hasToken`


This PR includes an optimization in 2 parts.

1. Implement the infrastructure to execute functions with access to a text index instead of the column
2. Optimize the ActionsDAG to substitute the original user function with a new one when possible.

As support (minor) changes this include:
a. Add extra information to the context in order to access from the function.
b. Cache index information to avoid extra unneeded multiple IO
c. New IFunction derived interface class that stores a reference to the context.
d. Code to translate from a roraring segments into row indexed array.

To use the new feature:

```sql
SET query_plan_direct_read_from_text_index = 1

SELECT count() FROM sometable WHERE hasToken(text_col, 'token')
SELECT sum(other_column) FROM sometable WHERE hasToken(text_col, 'token')
```

where `text_col` is a text column with a text index.

Ergus committed:
Start skeleton first steps

Ergus committed:
Add storage_snapshot information to context

ATM this is initialized in the ReadFromMergeTree constructor because
1. That constructor seems to always take place early enough
2. It is called from all the multiple places where we need them to be set.

Ergus committed:
Add index information to context.

The addition is lazy, because the indexes are initialized in a delayed moment after the ReadFromMergeTree constructor.
This is actually a problem for generalization, because is the call from `interpreter_with_analyzed->GetQueryPlan()`
To go around the issue we only attempt to use the index information from the context when input_rows_count > 0 which at the moment seems too
be a "good enough" solution

Ergus committed:
Substitute pointer with shared pointer for index in context

This solves the issue that the index are stored in ReadFromMergeTree class that is destroyed with the planner, before the pipeline functions
execute.
However, this is not a final solution because it duplicates the index class which could become a problem.
The final solution involves using the same shared pointer to avoid duplication, but that implies multiple modifications in the
ReadFromMergeTree class.
Remember that the shared_ptr is not thread safe, which is a problem and needs explicit handling in the destruction to ensure the proper
reference counter.
TODO: Once everything works, remember to implement this properly without supplications

Ergus committed:
Move getMergeTreeReaderSettings to a static function

* src/Storages/MergeTree/MergeTreeIOSettings.{h,cpp} (MergeTreeReaderSettings::Create) : New replacement function
Update all the occurrences
* src/Processors/QueryPlan/ReadFromMergeTree.cpp (getMergeTreeReaderSettings) : Removed function

Ergus committed:
Extend getSkippingIndices to also store and set range in context

TODO: Rename the function

Ergus committed:
Works!!!

Ergus committed:
Solve index issues and add extra checks

TODO: still searching the correctness issue

Ergus committed:
Fixed corner case issue.

Now the numbers are correct.

Ergus committed:
Second optimization (~5x)

Avoid re-reading the PostingsCacheForStore by caching the values in
the context.
It was already read during filterMarksUsingIndex.
I follow the same approach that with _skip_indexes here. Add a
shared_lock to hold and avoid race conditions due to parallel function
executions

Ergus committed:
Make UsefulSkipIndexes a shared_ptr

This change is to share the same reference in the reader and the
context, needed because the ReadFromMergeTree is destroyed before the
pipeline executes. Previously we created an extra copy instead.

Ergus committed:
Correct handling of context data

This is a big commit, but doesn't really change the main logic.

* src/Interpreters/IndexContextInfo.h (IndexContextInfo): New data
structure to store the index information that will be stored in the
context.
* src/Processors/QueryPlan/ReadFromMergeTree.cpp (BuildIndexes):
Doesn't modify the context anymore because it hasn't filtered all the
marks yet.
*
src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp (filterPartsByPrimaryKeyAndSkipInd)
: Use the new structure to store index information in the
context. This change simplifies the access to the context as the part
information is set in parallel in a temporal vector (lock-free) and
the ranges are copied after passing multiple filters, which is good to
avoid reading granules without matches.
* src/Functions/hasTokenIndex.h: Use the new context structure with
the index information. Add code to handle filtered/skipped marks more
efficiently.

Ergus committed:
Prepare code to implement granularity thing

Ergus committed:
Generalize implementation considering index granularity.

This implementation considers index granularity which by default is >
1.
In that case the attempt to read the index too many times will not
scale. SO, our solution implies calling the matches only once with the
ranges as arguments.
This approach is specially good when there are few ranges to match in
a single index with big granularity.

TODO: After a profile I expect to see more cost reading the index with
granularity = 64 (default) But we already have a solution for that ;)

In this commit I also generalize the implementation to use formal data
structures and remove some micro optimizations that didn't give any
significant performance improvement.

Ergus committed:
Cleanup comments

Ergus committed:
Fix match iteration

* src/Functions/hasTokenIndex.h (postingArrayToOutput) : Protect vs empty entries and out of bound matches.

Ergus committed:
New function to register columns changes in the reader

* src/Processors/QueryPlan/ReadFromMergeTree.{h,cpp} (registerColumnsChanges) : New function

Ergus committed:
Create a special class to identify optimizable functions

* src/Functions/FunctionsStringSearch.h (FunctionsStringSearchBase): This is a new class that is intended to contain all the information
needed to optimize a functions. At the moment it will only store a pointer to the context that needs to be passed to the optimized function.
In the future this function should provide all the optimized function information itself.

Ergus committed:
Implement function substitution as an optimization step in the fist pass.

* src/Processors/QueryPlan/Optimizations/useTextIndexSearch.cpp: Implement a helper wrapper class that performs the node substitution in the
dag. See the comments inlined in code for more information.
* src/Interpreters/ActionsDAG.h: Declare the helper class as a friend to modify the dag without making too much of a mess in the ActionsDAG
code.
Also to reduce the so undesirable "coupling"

TODO: At the moment the optimization works by creating a new function node and redirecting everything.

Ergus committed:
Generalize detection of IndexOptimized functions.

* src/Functions/FunctionsStringSearchBase.h (FunctionsStringSearchBase) : New file for the base class
* src/Functions/FunctionsStringSearch.h (FunctionsStringSearch) : Add new template argument to specify optimization information.
* src/Functions/hasToken.cpp : Specify optimization info to hasToken function.
* src/Functions/hasTokenIndex.h : Also derive from new base class.
* src/Storages/MergeTree/RPNBuilder.{h,cpp} (RPNBuilderFunctionTreeNode::isOptimizedIndexFunction) : New function to detect index optimized functions.
* src/Storages/MergeTree/MergeTreeIndexGin.cpp (traverseASTEqualsIndex) : Change checks to throw an exception.
(traverseAtomAST) : Use the new condition

TODO: At the moment we still use the contains function. That needs to change.

rschu1ze commented:
@Ergus Could you please merge from master? This PR is in bad bad merge conflicts :-/

shankar-iyer commented on src/Processors/QueryPlan/ReadFromMergeTree.cpp:
```diff
@@ -2859,4 +2830,41 @@ void ReadFromMergeTree::describeProjections(JSONBuilder::JSONMap & map) const
+        output_header = MergeTreeSelectProcessor::transformHeader(
+            storage_snapshot->getSampleBlockForColumns(all_column_names),
+            lazily_read_info, query_info.prewhere_info);
+    }
```
Check `updateLazilyReadInfo() `and `replaceVectorColumnWithDistanceColumn()`, need to add this code block to reflect `all_column_names` updates.

```
    /// if analysis has already been done (like in optimization for projections),
    /// then update columns to read in analysis result
    if (analyzed_result_ptr)
        analyzed_result_ptr->column_names_to_read = all_column_names;
```

shankar-iyer commented on src/Functions/hasTokenIndex.h:
```diff
@@ -0,0 +1,316 @@
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
```
Just for my understanding, does this function get called on a `index_granularity` sized bunch of rows or the bunch could be  less than or more than `granule` size?

shankar-iyer commented on src/Functions/hasTokenIndex.h:
```diff
@@ -0,0 +1,316 @@
+        const size_t marks_count = part->index_granularity->getMarksCountWithoutFinal();
+        const size_t index_marks_count = (marks_count + index_granularity - 1) / index_granularity;
+
+        MergeTreeIndexReader reader(
```
There is one interesting aspect to be handled. The text index could have been _defined_ on the table, but not yet _built_ / _materialized_ on some or all `parts`.  Is that handled currently?

shankar-iyer commented on src/Functions/hasTokenIndex.h:
```diff
@@ -0,0 +1,316 @@
+        for ( const auto &[index_mark, subranges] : ranges_map)
+        {
+            MergeTreeIndexGranulePtr granule = nullptr;
+            reader.read(index_mark, granule);
```
Just for my understanding - does the index has to be **read back again** here? Is the text index not in **some data structure in memory** when we reach here? Is it common for users to write queries like this -
```
SELECT ...
FROM ..
WHERE hasToken(message, 'INCORRECT_DATA') OR hasToken(message, 'BAD_ARGUMENTS')
```
Will we read the index twice (even if it is cached read)?

Ergus committed:
Merge remote-tracking branch 'principal/master' into full_text_search_optimization

rschu1ze reviewed (commented):
I took an initial look, didn't check lots of functions and touched files yet because I couldn't make this PR run locally due to the merge conflicts.

Next steps:
- get https://github.com/ClickHouse/ClickHouse/pull/85323 merged as soon as possible
- merge `master` into this PR
- turn it from draft statust to ready-for-review so CI runs on it

Ergus committed:
CR1 @rschu1ze

allow_experimental_text_index_pipeline -> query_plan_direct_read_from_text_index (in settings)
allow_experimental_text_index_pipeline -> direct_read_from_text_index (in structs)
tryUseTextSearch -> tryDirectReadFromTextIndex
isIndexedColumnNode -> isColumnNodeWithTextIndex
hasIndexedChild -> hasChildColumnNodeWithTextIndex

Ergus commented on src/Processors/QueryPlan/Optimizations/useTextIndexSearch.cpp:
```diff
@@ -0,0 +1,464 @@
+        chassert(map_it != map_indexed_columns.end());
+
+        { /// Check if the associated index column already exist
+            const auto index_column_it = std::find_if(dag.nodes.rbegin(), dag.nodes.rend(),
```
Because when we insert the index column it is closer to the end of the list.
So in the next substitution if the index is the same we can just look from the end. I know it is not critical at all, but IMO is a good hint with the cost of 3 r's.

Ergus commented on src/Processors/QueryPlan/Optimizations/useTextIndexSearch.cpp:
```diff
@@ -0,0 +1,464 @@
+        /// 2. Adding two extra arguments, the virtual columns: _part_index and _part_offset
+        ActionsDAG::NodeRawConstPtrs new_children;
+        new_children.reserve(original_function_node.children.size() + 2);
+        std::ranges::transform(
```
:(

Ergus committed:
CR1.1 @rschu1ze

Ergus committed:
Remove some prints used on debug

Generalize the condition in src/Storages/MergeTree/MergeTreeIndexGin.cpp
(this change is not needed but we will do a pertinent change associated in a while)

Ergus commented on src/Processors/QueryPlan/Optimizations/useTextIndexSearch.cpp:
```diff
@@ -0,0 +1,464 @@
+        if (replaced == 0)
+            return nullptr;
+
+        if (pindex == nullptr)
```
Comment added

Ergus committed:
CR1.2 @rschu1ze

hasTokenIndex -> _hasToken_index

Ergus committed:
Fix Style check issues

Ergus committed:
Fix Fast test issue

Ergus committed:
Merge remote-tracking branch 'origin/master' into full_text_search_optimization

Ergus committed:
Merge remote-tracking branch 'origin/master' into full_text_search_optimization

Due to another conflict

Ergus commented:
Waiting for #85323 again

ahmadov commented on src/Interpreters/Context.h:
```diff
@@ -1456,6 +1468,24 @@ class Context: public ContextData, public std::enable_shared_from_this<Context>
+        if (index_info)
+            return {index_info, std::move(slock)};
+
+        return {nullptr, std::shared_lock<std::shared_mutex>{}};
```
isn't this a case for an error? or in which case can this happen?

ahmadov commented on src/Interpreters/GinFilter.cpp:
```diff
@@ -217,6 +212,67 @@ size_t GinFilter::memoryUsageBytes() const
+    const PostingsCacheForStore *cache_store,
+    const MarkRanges &ranges) const
+{
+    chassert(ranges.size() > 0);
```
[minor]
```suggestion
    chassert(!ranges.empty());
```

ahmadov commented on src/Interpreters/Context.h:
```diff
@@ -1456,6 +1468,24 @@ class Context: public ContextData, public std::enable_shared_from_this<Context>
     std::pair<Context::StorageSnapshotCache *, std::unique_lock<std::mutex>> getStorageSnapshotCache() const;
 
+     /// Information needed in pure index functions (without real column access).
+    std::pair<IndexContextInfoPtr, std::shared_lock<std::shared_mutex>> getIndexInfo() const
```
Maybe add a simple comment that functions can access this info in parallel but only for reading purpose. Also, maybe return `std::pair<const IndexContextInfoPtr, ...` as another hint that it's not modifiable

ahmadov commented on src/Interpreters/GinFilter.cpp:
```diff
@@ -188,13 +189,7 @@ bool GinFilter::contains(const GinFilter & filter, PostingsCacheForStore & cache
-        postings_cache = reader.createPostingsCacheFromTerms(filter.getTerms());
-        cache_store.cache[filter.getQueryString()] = postings_cache;
-    }
+    GinPostingsCachePtr postings_cache = cache_store.getPostings(filter);
```
[minor] maybe `getOrRetrievePostings` as hint

ahmadov commented on src/Interpreters/GinFilter.cpp:
```diff
@@ -217,6 +212,67 @@ size_t GinFilter::memoryUsageBytes() const
+    const GinPostingsCachePtr postings_cache = cache_store->getCachedPostings(*filter);
+    GinIndexPostingsList posting_bitset;
+
+    for (const GinSegmentWithRowIdRange &range : rowid_ranges)
```
[minor]
```suggestion
    for (const GinSegmentWithRowIdRange & rowid_range : rowid_ranges)
```

ahmadov commented on src/Interpreters/GinFilter.cpp:
```diff
@@ -217,6 +212,67 @@ size_t GinFilter::memoryUsageBytes() const
+                || container_it->second->maximum() < range.range_start
+                || container_it->second->minimum() > range.range_end)
+            {
+                range_matches.removeRange(range.range_start, range.range_end + 1);
```
looking at the removeRange from CRoaring, it seems it's bit complex and probably costly. Can we just somehow skip it or initialize ranges_matches empty and use `OR (|)` afterwards for the matching range?

ahmadov commented on src/Interpreters/GinFilter.cpp:
```diff
@@ -217,6 +212,67 @@ size_t GinFilter::memoryUsageBytes() const
+    range_bitset &= posting_bitset;
+
+    const size_t cardinality = range_bitset.cardinality();
+    std::vector<uint32_t> indices(cardinality);
```
do we need to extract them to an array? Can we pass the roaring bitmap?

ahmadov commented on src/Interpreters/IndexContextInfo.h:
```diff
@@ -0,0 +1,36 @@
+};
+
+
+using IndexContextInfoPtr = std::shared_ptr<IndexContextInfo>;
```
duplicate of https://github.com/ClickHouse/ClickHouse/blob/c8ce2c0c6374fc19e3aba61f3018bf183588e533/src/Interpreters/Context.h#L575, can we reuse it?

ahmadov commented on src/Processors/QueryPlan/Optimizations/optimizeTree.cpp:
```diff
@@ -8,6 +8,11 @@
 
 #include <stack>
 
+#include <cstdio>
```
leftovers?

ahmadov commented on src/Processors/QueryPlan/ReadFromMergeTree.cpp:
```diff
@@ -1780,7 +1748,7 @@ static void buildIndexes(
     }
 
-    UsefulSkipIndexes skip_indexes;
+    std::shared_ptr<UsefulSkipIndexes> skip_indexes = std::make_shared<UsefulSkipIndexes>();
```
what is the reason for changing it into `std::shared_ptr<>`?

ahmadov reviewed (commented):
Not finished it yet, I haven't checked the main parts yet.

Ergus commented on src/Functions/hasTokenIndex.h:
```diff
@@ -0,0 +1,315 @@
+
+
+template <FunctionIndexConcept Impl>
+class FunctionIndex : public FunctionsStringSearchBase
```
Yes, the name is bad, but I plan to use the same code for the other functions, so maybe we need something not so specific... what about `FunctionSearchTextIndex` or something like that?

ahmadov replied:
yeah `FunctionSearchTextIndex` or `FunctionTextIndexTokenSearch`  would be good, it's up to you :)

Ergus committed:
CR2.0 @ahmadov

Ergus committed:
Merge remote-tracking branch 'origin/master' into full_text_search_optimization

Ergus committed:
Fix duplication in dag transformation nodes list

* src/Processors/QueryPlan/Optimizations/useTextIndexSearch.cpp (tryGetOrCreateIndexColumn) : Format the function name to search.

Ergus committed:
Reflect all_column_names updates

* src/Processors/QueryPlan/ReadFromMergeTree.cpp (ReadFromMergeTree::registerColumnsChanges) : Set column_names_to_read  in analyzer result

Ergus committed:
Merge remote-tracking branch 'principal/master' into full_text_search_optimization

Ergus committed:
Small improvement to _hasToken_Index subranges.

Use more restricted subranges

Ergus committed:
Use inplace Dag function replacement.

This simplifies a lot of code by replacing the node content inplace.
Ad the node address is the same we can avoid the full redirection step and replacement we used to do.
However there is a small hack because we add a new node to pop it immediately, but so far that is not an issue.

Ergus committed:
Improve getIndices by avoiding a temparal extra roaring map.

This is a test at the moment and still needs improvement.

Ergus committed:
Fix and simplify postingArrayToOutput.

This is also slightly faster.

Ergus committed:
Check index materialization before dag substitution.

This is a preliminar solution that needs improvement

Ergus committed:
Merge remote-tracking branch 'principal/master' into full_text_search_optimization

Ergus committed:
Fix corner case with index boundaries.

This fixes the introduction of out of boundaries values for filtering in the index function.
The code is totally defensive now by using a proper number of assertions and introducing the tight ranges for filtering.
Any out of bound issue will be handled properly in release code (and produce correct results), but will trigger some assertion in debug
mode.

rschu1ze committed:
Merge remote-tracking branch 'ClickHouse/master' into full_text_search_optimization

rschu1ze committed:
Fix clang-tidy, style check and FastTest

rschu1ze committed:
Enable direct read optimization

rschu1ze committed:
Cosmetics

rschu1ze committed:
Cosmetics, pt. II

rschu1ze committed:
Cosmetics

rschu1ze committed:
Cosmetics

rschu1ze committed:
Move some code around

Ergus committed:
Remove deprecated optimization condition

Check the function->info instead of the existence of context.

rschu1ze committed:
Cosmetics

rschu1ze committed:
Cosmetics

Ergus committed:
Cleanup

* src/Processors/QueryPlan/Optimizations/useDirectReadFromTextIndex.cpp (printDag) :
(printHierarchicalActions) : Removed

rschu1ze committed:
Merge remote-tracking branch 'ClickHouse/master' into full_text_search_optimization

rschu1ze committed:
Minor fixups

rschu1ze committed:
Optimize includes

rschu1ze committed:
Cosmetics

rschu1ze committed:
Improve separation of concerns

rschu1ze committed:
Less ambiguous naming

rschu1ze committed:
Cleanup hasTokenIndex docs

rschu1ze committed:
Cosmetics

rschu1ze committed:
Cosmetics

rschu1ze committed:
Kill needless complexity

rschu1ze committed:
Rid ourselves of unneeded compexity

rschu1ze committed:
Fix clang-tidy

rschu1ze committed:
Move some code around

rschu1ze committed:
Micro code beautification

Ergus committed:
Use named argument var

Ergus committed:
Fix incorrect/unneeded assertion

Ergus committed:
Merge remote-tracking branch 'principal/master' into full_text_search_optimization

Ergus committed:
Improve node replacement condition

The number of replaced functions is different from removed functions.

We should use that to condition `read_from_mergetree_step->registerColumnsChanges` otherwise we will be adding the virtual columns even when
not needed.

Ergus committed:
Condition setting context index information.

Don't pay for what you don't use

Ergus committed:
Merge remote-tracking branch 'principal/master' into full_text_search_optimization

helifu commented:
Sorry, is this PR related to the [private preview version](https://clickhouse.com/blog/clickhouse-full-text-search)?

rschu1ze commented:
> Sorry, is this PR related to the [private preview version](https://clickhouse.com/blog/clickhouse-full-text-search)?

Kind of. The PR speeds up certain common text search queries.

Ergus committed:
Fix incorrect option access.

I would expect that this never happens because the external filtering should exclude it.
But it seems like it does in the CI even when I cannot reproduce locally.

rschu1ze commented:
>Anyway, the full-text search description in [the blog](https://clickhouse.com/blog/clickhouse-full-text-search) is open source or close? 😁

It's open source.

Ergus committed:
Merge remote-tracking branch 'principal/master' into full_text_search_optimization

Ergus committed:
Partially revert the needsIndexInfo optimization

I will push a better approach in a while

Ergus committed:
Merge remote-tracking branch 'principal/master' into full_text_search_optimization

rschu1ze commented:
Will be part of #86485
